File: Lab 12/src/TextEditor.java
public class TextEditor {
    private String content;
    private int cursorPosition;
    private String selection;
    private final History history;

    public TextEditor() {
        content = "";
        cursorPosition = 0;
        selection = "";
        history = new History(); // Initialize the history to keep track of mementos
    }

    // Method to set content (overwrite existing content)
    public void setContent(String content) {
        history.backup(this); // Backup the current state before setting new content
        this.content = content; // Set the content to the new value
        cursorPosition = content.length(); // Update cursor position
    }

    // Method to write text to the editor (append text to content)
    public void write(String text) {
        history.backup(this); // Backup the current state before writing
        content += text; // Append new text to content
        cursorPosition = content.length(); // Move cursor to the end of the content
    }

    // Undo the last write or content set operation
    public void undo() {
        history.undo(this); // Restore the previous state using History
    }

    // Getter method to get the content
    public String getContent() {
        return content;
    }

    // Create a Memento to store the current state of the editor
    public EditorMemento createMemento() {
        return new EditorMemento(content, cursorPosition, selection);
    }

    // Restore the state from a Memento
    public void restoreFromMemento(EditorMemento memento) {
        this.content = memento.content();
        this.cursorPosition = memento.cursorPosition();
        this.selection = memento.selection();
    }
}

File: Lab 12/TextEditor.cs
public class Texteditor
{
    private string _content;
    private int _cursorposition;
    private List<string> _selections;

    public Texteditor()
    {
        _content = string.Empty;
        _cursorPosition = 0;
        _selections = new List<string>();
    }

    public void SetContent(string content)
    {
        _content = content;
    }
    public SetCursorPosition(int position)
    {
        if(position < 0 || position > _content.Length)
        {
            Console.WriteLine("Invalid cursor position.");
            return;
        }
        _cursorposition=position;
    }
    public AddSelection(string selection)
    {
        _selections.Add(selection);
    }
    public void ClearSelections()
    {
        _selections.Clear();

    }
    public void DisplayState()
    {
        Console.WriteLine($"Content: {_content}");
        Console.WriteLine($"Cursor Position: {_cursorPosition}");
        Console.WriteLine("Selections: " + string.Join(", ", _selections));
    }
     public void CreateMemento(){
        return new EditorMemento(_content, _cursorPosition, _selections);
    }
    public void RestoreFromMemento(EditorMemento memento){
        _content=memento.Content;
        _cursorPosition=memento.CursorPosition;
        _selections=new List<string>(memento.Selections);
    }
}